test/util/malicious/hasher.go

```diff
@@ -0,0 +1,239 @@
+package malicious
+
+import (
+	"bytes"
+	"errors"
+	"hash"
+
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+// NOTE: This file is a copy of the original nmt.Hasher implementation, but with
+// all of the validation logic removed. This allows for malicious nodes to create nmt roots that are out of order.
+
+const (
+	LeafPrefix = 0
+	NodePrefix = 1
+)
+
+var _ hash.Hash = (*Hasher)(nil)
+
+var (
+	ErrUnorderedSiblings         = errors.New("NMT sibling nodes should be ordered lexicographically by namespace IDs")
+	ErrInvalidNodeLen            = errors.New("invalid NMT node size")
+	ErrInvalidLeafLen            = errors.New("invalid NMT leaf size")
+	ErrInvalidNodeNamespaceOrder = errors.New("invalid NMT node namespace order")
+)
+
+type Hasher struct {
```

Comment:
had to copy paste fork this from nmt in order to remove all of the protections against leafs that are out of order

Reply:
> As a note, it seems outside the scope of a hasher to have all these protections in place. That feels like it belongs to the jurisdiction of the tree itself

Just a note on this, there are two hashers: one is the base hasher e.g., sha256, and one is the nmt hasher, or the namespace hasher, that utilizes the base hasher. The latter actually is aware of the format of the leaves and the nodes i.e., their min and max namespace, their ordering. Due to this, all the functions associated with the namespace hasher validate the format of inputs accordingly. However, the former, i.e., the basehasher, is completely unaware and is only responsible for the digest calculation.  I am not sure, but I think you might be referring to the base-hasher, which, as you suggested, should be and is clueless about the namespace-related logic.